### PR TITLE
[Pending - Wait add new test & doc] Add .withTransaction(int: IsolationLevel)

### LINKED ISSUE #8466: Allow to set the transaction isolation level
At the moment, there is no way to specify the transaction isolation level for a single transaction, it would be great to have a syntax like this:
 ```scala
database.withTransaction(TransactionIsolationLevel.Serializable) { implicit conn => ... }
```

The code changes to support that are trivial, would you accept a pull request?

### PR DESCRIPTION
**Pull request by nuboat**
# Pull Request Checklist

* [X] Have you read [How to write the perfect pull request](https://github.com/blog/1943-how-to-write-the-perfect-pull-request)?
* [X] Have you read through the [contributor guidelines](https://www.playframework.com/contributing)?
* [X] Have you signed the [Lightbend CLA](https://www.lightbend.com/contribute/cla)?
* [X] Have you referenced any issues you're fixing using [commit message keywords](https://help.github.com/articles/closing-issues-using-keywords/)?
* [X] Have you added copyright headers to new files?
* [X] Have you checked that both Scala and Java APIs are updated?
* [ ] Have you updated the documentation for both Scala and Java sections?
* [ ] Have you added tests for any changed functionality?

# Helpful things

## Fixes

Fixes #8466 

## Purpose

Allow set Transaction Isolation level

## Background Context

I love Playframework and need to join Hacktoberfest

## References

Are there any relevant issues / PRs / mailing lists discussions?

### CODE REVIEW

File: framework/src/play-java-jdbc/src/main/java/play/db/DefaultDatabase.java [hunk 1]
```diff
@@ -21,7 +21,7 @@
 
     private final play.api.db.Database db;
 
-    public DefaultDatabase(play.api.db.Database database) {
+    public DefaultDatabase(final play.api.db.Database database) {
```
Comment (AlexITC) [inline framework/src/play-java-jdbc/src/main/java/play/db/DefaultDatabase.java]: This is not related to the PR.

File: framework/src/play-java-jdbc/src/test/java/play/db/DatabaseTest.java [hunk 2]
```diff
@@ -144,6 +144,26 @@ public void provideConnectionHelpers() {
         db.shutdown();
     }
 
+    @Test
+    public void provideConnectionHelpers2() {
```
Comment (AlexITC) [inline framework/src/play-java-jdbc/src/test/java/play/db/DatabaseTest.java]: Could we use a more specific name? in any case, most of the code could be reused, you might want to create a common method.

File: framework/src/play-jdbc-api/src/main/java/play/db/Database.java [hunk 3]
```diff
@@ -174,6 +199,11 @@ public Connection getConnection(boolean autocommit) {
                 return Database.this.withTransaction(block::apply);
             }
 
+            public <A> A withTransaction(final Enumeration.Value isolationLevel,
+                                         final scala.Function1<Connection, A> block) {
+//TODO
```
Comment (AlexITC) [inline framework/src/play-jdbc-api/src/main/java/play/db/Database.java]: TODO, what?

File: framework/src/play-jdbc-api/src/main/java/play/db/TransactionIsolationLevel.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+    ReadCommited(Connection.TRANSACTION_READ_COMMITTED),
+
+    RepeatedRead(Connection.TRANSACTION_REPEATABLE_READ),
+
+    Serializable(Connection.TRANSACTION_SERIALIZABLE);
+
+    private final int id;
+
+    TransactionIsolationLevel(final int id) {
```
Comment (AlexITC) [inline framework/src/play-jdbc-api/src/main/java/play/db/TransactionIsolationLevel.java]: What prevents me from creating a `TransactionIsolationLevel` with a wrong id?

File: framework/src/play-jdbc-api/src/main/scala/play/api/db/Database.scala [hunk 5]
```diff
@@ -5,7 +5,9 @@
 package play.api.db
 
 import java.sql.Connection
+
 import javax.sql.DataSource
+import play.api.db.TransactionIsolationLevel.TransactionIsolationLevel
```
Comment (AlexITC) [inline framework/src/play-jdbc-api/src/main/scala/play/api/db/Database.scala]: This looks weird, don't you think?

File: framework/src/play-jdbc/src/test/scala/play/api/db/DatabasesSpec.scala [hunk 6]
```diff
@@ -112,7 +112,7 @@ class DatabasesSpec extends Specification {
         results.next must beFalse
       }
 
-      db.withTransaction { c =>
+      db.withTransaction(TransactionIsolationLevel.Serializable) { c =>
```
Comment (AlexITC) [inline framework/src/play-jdbc/src/test/scala/play/api/db/DatabasesSpec.scala]: Instead of updating a single test, I would leave this intact and create a test per isolation level, probably with some corner cases that ensure the isolation level is working as expected.

File: framework/src/play-java-jdbc/src/test/java/play/db/DatabaseTest.java (hunk 2 above)
Comment (nuboat) [reply framework/src/play-java-jdbc/src/test/java/play/db/DatabaseTest.java]: This is not related to PR, I found a project missing some test case (withConnection(autocommit = false... ). Do you want me to keep this?

File: framework/src/play-jdbc-api/src/main/java/play/db/Database.java (hunk 3 above)
Comment (nuboat) [reply framework/src/play-jdbc-api/src/main/java/play/db/Database.java]: Oops, I miss some part.

File: framework/src/play-jdbc-api/src/main/java/play/db/TransactionIsolationLevel.java (hunk 4 above)
Comment (nuboat) [reply framework/src/play-jdbc-api/src/main/java/play/db/TransactionIsolationLevel.java]: This cannot happen.
"If no access modifier is specified for the constructor of an enum type, the constructor is private".

File: framework/src/play-jdbc-api/src/main/scala/play/api/db/TransactionIsolationLevel.scala [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+ */
+object TransactionIsolationLevel extends Enumeration {
+  type TransactionIsolationLevel = Value
+
+  val ReadUncommitted: Value = Value(Connection.TRANSACTION_READ_UNCOMMITTED)
+  val ReadCommited: Value = Value(Connection.TRANSACTION_READ_COMMITTED)
+  val RepeatedRead: Value = Value(Connection.TRANSACTION_REPEATABLE_READ)
+  val Serializable: Value = Value(Connection.TRANSACTION_SERIALIZABLE)
+
+  //sealed abstract class TransactionIsolationLevel(val id: Int)
```
Comment (gmethvin) [inline framework/src/play-jdbc-api/src/main/scala/play/api/db/TransactionIsolationLevel.scala]: Why use the enumeration rather than this?

Comment (nuboat) [reply framework/src/play-jdbc-api/src/main/scala/play/api/db/TransactionIsolationLevel.scala]: I have no idea what it's should be on first time.  In case we prefer use  class than enum, I will push the update.

Comment (nuboat) [reply framework/src/play-jdbc-api/src/main/scala/play/api/db/TransactionIsolationLevel.scala]: I have no idea what it's should be on first time.  In case we prefer use  class than enum, I will push the update.

Comment (nuboat) [reply framework/src/play-jdbc-api/src/main/scala/play/api/db/TransactionIsolationLevel.scala]: I have no idea what it's should be on first time.  In case we prefer use  class than enum, I will push the update.

Comment (nuboat) [reply framework/src/play-jdbc-api/src/main/scala/play/api/db/TransactionIsolationLevel.scala]: So, Should I change to "sealed abstract class TransactionIsolationLevel(val id: Int)"?

Comment (AlexITC) [reply framework/src/play-jdbc-api/src/main/scala/play/api/db/TransactionIsolationLevel.scala]: If it is compatible with Java, it is far better to use the ADT.

Comment (gmethvin) [reply framework/src/play-jdbc-api/src/main/scala/play/api/db/TransactionIsolationLevel.scala]: The `sealed abstract class` is preferred.

File: framework/src/play-jdbc/src/test/scala/play/api/db/DatabasesSpec.scala [hunk 8]
```diff
@@ -125,6 +125,15 @@ class DatabasesSpec extends Specification {
       }
     }
 
+    "manual setup trasaction isolation level" in new WithDatabase {
+      val db = Databases.inMemory(name = "test-manualSetupTrasactionIsolationLevel")
+
+      db.withTransaction(TransactionIsolationLevel.Serializable) { c =>
+        c.createStatement.execute("create table test (id bigint not null, name varchar(255))")
+        c.createStatement.execute("insert into test (id, name) values (1, 'alice')")
```
Comment (cchantep) [inline framework/src/play-jdbc/src/test/scala/play/api/db/DatabasesSpec.scala]: would test `c.getTransactionIsolation` is what's expected

File: framework/src/play-jdbc-api/src/main/scala/play/api/db/TransactionIsolationLevel.scala (hunk 7 above)
Comment (marcospereira) [reply framework/src/play-jdbc-api/src/main/scala/play/api/db/TransactionIsolationLevel.scala]: Do we actually need Scala and Java versions of this? Why not reuse the Java enum?

**Comment by marcospereira** [comment]
Overall, this looks good to me. I just left a comment that maybe we don't need `TransactionIsolationLevel` defined twice in Scala and Java. I think the Java enum is enough and we can reuse in Scala code.

WDYT?

File: framework/src/play-jdbc-api/src/main/scala/play/api/db/TransactionIsolationLevel.scala (hunk 7 above)
Comment (gmethvin) [reply framework/src/play-jdbc-api/src/main/scala/play/api/db/TransactionIsolationLevel.scala]: If it were a Java enum then that'd be ok, but these are actually just `int`s. Scala people like type safety so I think it's justified to use the ADT.

Comment (gmethvin) [reply framework/src/play-jdbc-api/src/main/scala/play/api/db/TransactionIsolationLevel.scala]: Oh, I didn't notice there we had defined an enum in Java. Ignore that comment.

**Comment by gmethvin** [comment]
I think the main argument for using the Scala ADT is to match the idiomatic style in Scala, and for consistency with the rest of Play's APIs. I know some people also exclude the Play Java API packages from auto-import in their IDEs, so it's slightly more cumbersome to use a Java API class.

File: framework/project/BuildSettings.scala [hunk 9]
```diff
@@ -671,7 +671,13 @@ object BuildSettings {
       ProblemFilters.exclude[IncompatibleResultTypeProblem]("play.api.test.Helpers.stubMessagesApi$default$6"),
       ProblemFilters.exclude[DirectMissingMethodProblem]("play.api.test.Helpers.stubMessagesApi"),
       ProblemFilters.exclude[IncompatibleResultTypeProblem]("play.api.test.StubMessagesFactory.stubMessagesApi$default$6"),
-      ProblemFilters.exclude[DirectMissingMethodProblem]("play.api.test.StubMessagesFactory.stubMessagesApi")
+      ProblemFilters.exclude[DirectMissingMethodProblem]("play.api.test.StubMessagesFactory.stubMessagesApi"),
+
+      // Add play.db.Database.withTransaction config
+      ProblemFilters.exclude[ReversedMissingMethodProblem]("play.db.Database.withTransaction"),
+      ProblemFilters.exclude[ReversedMissingMethodProblem]("play.db.Database.withTransaction"),
```
Comment (PromanSEW) [inline framework/project/BuildSettings.scala]: Duplicated line

**Comment by mkurz** [comment]
Needs to be backported to 2.7.x

**Comment by marcospereira** [comment]
Backport to 2.7.x: 99b338fb6
